### ISSUE TITLE
Can't recover from invalid markup in code editor 

### ISSUE THREAD

**Issue by mkaz** [WordPress/gutenberg#29988]
## Description

When pasting invalid markup into the Code Editor, in this case some HTML that does not have the proper closing `</code>` tag the attempt to validate the markup ends up creating an infinite loop of `<code></code>` tags.

## Step-by-step reproduction instructions

1. Create a new post
2. Switch to the Code Editor
3. Paste in the following markup
 
```
<ul>
  <li> Before </li>
  <li>
    <code>
    	Missing end tag
        
  </li>
  <li> After </li>
</ul>
```
4. Switch back to code editor to see what is happening.

## Expected behaviour

The editor needs to invalidate the markup properly

## Screenshots or screen recording (optional)

![code-paste-test](https://user-images.githubusercontent.com/45363/111666402-ec292980-87d0-11eb-8532-309ade5d6070.gif)

**Comment by Quintis1212** [comment]
Hi ) I will fix this bug )

**Comment by avivalk** [comment]
> Hi ) I will fix this bug )

@Quintis1212 will your fix also fix this issue? https://github.com/WordPress/gutenberg/issues/29338

When will the new update come out?

**Comment by Quintis1212** [comment]
Hi @avivalk  ) My fix do not solve yours #29338 issue right now , but I working on it and I have some thots how to fix #29338 issue.
New update will come after my PR will be merged - https://github.com/WordPress/gutenberg/pull/30313 , if you want you can clone my branch with my fix and build gutenberg editor that will have my fix

**Comment by avivalk** [comment]
> Hi @avivalk ) My fix do not solve yours #29338 issue right now , but I working on it and I have some thots how to fix #29338 issue.
> New update will come after my PR will be merged - #30313 , if you want you can clone my branch with my fix and build gutenberg editor that will have my fix

ok thank you very much

**Comment by mcsf** [comment]
I followed a [similar report](https://wordpress.org/support/topic/code-auto-generating-in-code-editor-after-5-7-update/), and I've tracked this issue down to #27717.

In this case, the offending markup to insert in the Code Editor can be boiled down to:

```html
<a href="1"><div></div></a><a href="2"></a>
```

The problem is threefold:

1. The change introduced in #27717 forces the Code Editor to re-parse and re-save its `value` every time `value` changes. This opens the door to infinite loops if there is a chance that parsing the post's content — in any of its stages — returns a different value.

2. If we are dealing with content that was wrapped in a Classic block, or if we are manually adding HTML in the Code Editor outside of the boundaries of a block (e.g. `<!-- wp:core/paragraph -->`), then — for backwards compatibility — the parser must run the HTML through [`autop`](https://github.com/WordPress/gutenberg/blob/trunk/packages/autop/README.md). This function is a faithful port of WordPress's original PHP [`wpautop`](https://developer.wordpress.org/reference/functions/wpautop/), and thus is compelled to break lines around `<div>` elements:

```html
INPUT:
<a href="1"><div></div></a><a href="2"></a>

OUTPUT:
<p><a href="1"></p>
<div></div>
<p></a>
<a href="2"></a></p>
```

this is because `<div>` elements are not [phrasing content](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#phrasing_content). If you replace `div` with `span` in the example, the bug does not occur. `autop` is behaving according to its specification, assuming the input isn't malformed and isolating the HTML in blocks of `<p>` and `<div>`.

3. When we delegate [attribute sourcing](https://github.com/WordPress/gutenberg/blob/e12881c79441ca874fb2b2c2adffba8ed0792103/packages/blocks/src/api/parser.js#L260) to the browser [via `hpq`](https://github.com/aduth/hpq/blob/master/src/index.js), the post's HTML is run through a blank HTML document. In essence:

```js
doc = document.implementation.createHTMLDocument('')
doc.body.innerHTML = postContent
return doc.body.innerHTML
```

This is handy, because the parsing and serialisation of HTML is handled natively by the browser through the `Document` interface, and the browser decides how to clean up any malformed HTML. Notice, however, how it handles our offending input. It closes the anchor at the end of the first paragraph, pulling the div out of the anchor, but for some reason it leaves an identical anchor **inside** the div:

```js
doc.body.innerHTML = '<p><a href="1"></p><div>oops</div><p></a><a href="2"></a></p>'
console.log(doc.body.innerHTML)
// prints:
// <p><a href="1"></a></p><div><a href="1">oops</a></div><p><a href="2"></a></p>
```

***

Given problems 2 (autop) and 3 (HTML parsing), we no longer have idempotence:

<img width="348" alt="Captura de ecrã 2021-04-05, às 19 10 37" src="https://user-images.githubusercontent.com/150562/113608532-a084d580-9642-11eb-8444-c0cf244e5a18.png">

Combined with problem 1 (automated saving behaviour), we have an infinite loop.

**Comment by mcsf** [comment]
This is now affecting WordPress 5.7. Some reports:

* https://wordpress.org/support/topic/font-tags-automatically-generate-while-editing/
* https://wordpress.org/support/topic/code-auto-generating-in-code-editor-after-5-7-update/

I'm tempted to **recommend reverting #27717**, at least for the purpose of patching 5.7. It's not a common bug, but when it manifests itself it tends to do so:

* in a very confusing way for the user, who may feel forced to close the editor, thereby risking data loss;
* when handling freeform/Classic content, thus in a context of backwards compatibility and handling old data.

Data integrity is important. Weighing our options, I'd rather have bug #9512 open (the one which #27717 sought to fix) than this one. Thoughts?

**Comment by noisysocks** [comment]
Is there a way to fix both this and #9512? cc. @grzim 

If it's not easy to do this, then I'm OK with reverting https://github.com/WordPress/gutenberg/pull/27717 for 5.7.x.

**Comment by grzim** [comment]
I think it would be best to revert the https://github.com/WordPress/gutenberg/pull/27717 and reopen https://github.com/WordPress/gutenberg/issues/9512 . The issue can be solved in a different way than saving content on blur, probably just on switching editors. Needs more investigation though.

**Comment by mkaz** [comment]
Closing as this issue was resolved with the revert to #27717
